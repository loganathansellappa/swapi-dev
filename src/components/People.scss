.PeopleContainer {
  display: flex;
  flex-direction: column;
  height: 100%;
  padding: 10px;
  .searchInput {
    width: 100%;
    background-image: linear-gradient(0deg, red 0%, blue 100%);
    color: #ffffff;
    font-size: 2rem;
    padding: 12px 20px;
    margin: 8px 0;
    display: inline-block;
    border: 1px solid #ccc;
    border-radius: 4px;
    box-sizing: border-box;
  }
  .people {
    display: flex;
    flex-wrap: wrap;
    padding-left: 10%;
    padding-top: 1%;
    padding-right: 10%;
    justify-content: center;
  }
  .paginate {
    display: flex;
    flex-wrap: nowrap;
    justify-content: center;
    margin-top: 10%;
  }
}
