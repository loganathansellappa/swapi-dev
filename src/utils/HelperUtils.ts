export const apiService = () => `${process.env.SWAP_URL}`;
export const capitalize = (s = 'n/a') => s[0].toUpperCase() + s.slice(1);
