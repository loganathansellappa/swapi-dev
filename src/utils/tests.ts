import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';

export const renderWithRouter = (ui: JSX.Element, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};
export const testStore = () => {
  const mockStore = configureStore();
  const initialState = { favoritePeople: { favourites: [] } };
  const store = mockStore(initialState);
  return store;
};

export const mockResource = () => {
  return {
    name: 'Luke Skywalker',
    height: '172',
    mass: '77',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    gender: 'male',
    homeworld: 'https://swapi.dev/api/planets/1/',
    films: [
      'https://swapi.dev/api/films/1/',
      'https://swapi.dev/api/films/2/',
      'https://swapi.dev/api/films/3/',
      'https://swapi.dev/api/films/6/',
    ],
    species: [],
    vehicles: [
      'https://swapi.dev/api/vehicles/14/',
      'https://swapi.dev/api/vehicles/30/',
    ],
    starships: [
      'https://swapi.dev/api/starships/12/',
      'https://swapi.dev/api/starships/22/',
    ],
    created: '2014-12-09T13:50:51.644000Z',
    edited: '2014-12-20T21:17:56.891000Z',
    url: 'https://swapi.dev/api/people/1/',
  };
};
