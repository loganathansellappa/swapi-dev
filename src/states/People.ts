import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { PersonType, ResourceType } from '../@types/People';

export interface PeopleState {
  favourites: Array<ResourceType> | [];
}

const initialState: PeopleState = {
  favourites: [],
};

export const peopleSlice = createSlice({
  name: 'people',
  initialState,
  reducers: {
    add: (state, action: PayloadAction<ResourceType>) => {
      state.favourites = [...state.favourites, action.payload];
    },
    remove: (state, action: PayloadAction<string>) => {
      state.favourites = state.favourites.filter(
        (person) => action.payload !== person.url
      );
    },
  },
});

export const { add, remove } = peopleSlice.actions;

export default peopleSlice.reducer;
