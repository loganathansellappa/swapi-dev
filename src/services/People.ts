import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import {ResourceResponse, ResourceType} from "../@types/People";
import {apiService} from "../utils/HelperUtils";

export const peopleApi = createApi({
    reducerPath: 'peopleApi',
    baseQuery: fetchBaseQuery({ baseUrl: apiService() }),
    endpoints: (builder) => ({
        getResource: builder.query<ResourceResponse, string>({
            query: (id) => `${id}`,
        }),
    }),
})

export const {useGetResourceQuery } = peopleApi