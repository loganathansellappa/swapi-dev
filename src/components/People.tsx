import React, {useMemo, useState} from "react";
import {ResourceResponse, ResourceType} from "../@types/People";
import {Person} from "./Person";
import Error from "./Error";
import {useGetResourceQuery} from '../services/People'
import {Loader} from "./Loader";
import "./People.scss";


const People: React.FunctionComponent  = () => {
    const [page, setPage] = useState<number>(1);
    const [search, setSearch] = useState('');
    const [data, setData] = useState<ResourceResponse | null>(null);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState(false);

    const updateSearch = (val: string) => {
        setPage(1)
        setSearch(val)
    }

    const peopleData =  useGetResourceQuery(`people?search=${search}&page=${page}`)
    const planetsData =  useGetResourceQuery(`planets?search=${search}&page=${page}`, {skip: !search.length})
    const shipData = useGetResourceQuery(`starships?search=${search}&page=${page}`, {skip: !search.length})

    useMemo(() => {
        let responses: ResourceResponse = {
            count: 0,
            next: null,
            previous: null,
            results: []
        }
        let validData = search.length ?  [peopleData, planetsData, shipData] : [peopleData]
        for (const { currentData:data, error:err, isLoading, isFetching } of validData) {
            console.log(err, 'err')
            if(isLoading || isFetching && !loading) {
                setLoading(true)
            }
            if(err && !error) {
                setLoading(error)
            }
            if(data) {
                if(data?.next) {
                    responses.next = data.next
                }
                setLoading(false)
                setError(false)
                responses.results = [...responses.results, ...data.results]
            }
        }
        setData(responses as ResourceResponse)
    }, [peopleData.data, planetsData.data, shipData.data])

    if (error) {return <Error />}

    return (
        <div className='PeopleContainer'>
            <div className="search">
                <input className={'searchInput'} value={search} placeholder="Search People/Planet/StarShips.." onChange={(e) =>  updateSearch((e.target as HTMLInputElement).value)}/>
            </div>
            {
                loading ?  <Loader size={100} className={'inline'}/> :
                <div className="people">
                    {data?.results?.map((person: ResourceType) => (
                        <Person key={person.url} person={person} />
                    ))}
                </div>
            }
                <div className="paginate">
                    <button  onClick={() => setPage((old) => Math.max(old - 1, 1))}  disabled={page === 1}>
                        {loading ? <Loader size={10} /> : 'Previous'}
                    </button>
                    <button onClick={() => setPage((old) => old + 1)} disabled={!data?.next} >
                        {loading ? <Loader size={10} />: 'Next'}
                    </button>
                </div>
        </div>
    );
};

export default People;
