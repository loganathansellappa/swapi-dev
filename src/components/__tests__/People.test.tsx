import { renderWithRouter } from '../..//utils/tests';
import People from "../People";

test('renders People Page', () => {
    const { container } = renderWithRouter(<People />);
    expect(container.firstChild).toMatchSnapshot();
});
