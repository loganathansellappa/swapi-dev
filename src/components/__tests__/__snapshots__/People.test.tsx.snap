// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person renders Person component 1`] = `
<div
  class="PeopleContainer"
>
  <div
    class="search"
  >
    <input
      class="searchInput"
      placeholder="Search People/Planet/StarShips.."
      value=""
    />
  </div>
  <div
    class="people"
  />
  <div
    class="paginate"
  >
    <button
      disabled=""
    >
      Previous
    </button>
    <button
      disabled=""
    >
      Next
    </button>
  </div>
</div>
`;
